### ISSUE TITLE
CI is stuck running integration tests for almost two hours

### ISSUE THREAD

**Issue by DnPlas** [canonical/kfp-operators#549]
### Bug Description

Integration tests in the CI are failing to be completed; initially the `bundle-integration-v2` job appears to be stuck and the only logs we can see are:

```
bundle-integration-v2: commands[0]> pytest -vv --tb=native -s --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2 /home/runner/work/kfp-operators/kfp-operators/tests/integration/test_kfp_functional_v2.py
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-8.3.2, pluggy-1.5.0 -- /home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/bin/python
cachedir: .tox/bundle-integration-v2/.pytest_cache
rootdir: /home/runner/work/kfp-operators/kfp-operators
configfile: pyproject.toml
plugins: operator-0.35.0, asyncio-0.21.2, anyio-4.4.0
asyncio: mode=strict
collecting ... collected 6 items
```

After almost two hours, the job fails and we get to see more logs. See from [L22](https://github.com/canonical/kfp-operators/actions/runs/10110429488/job/27960326357?pr=548#step:5:23).

Everything looks normal, the test cases are passing up until the ` tests/integration/test_kfp_functional_v2.py::test_viz_server_healthcheck` test case, which fails, BUT the logs from the failure do not indicate an error from `kfp-operators` or any other charm, what we see is:

```
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/connection.py", line 649, in rpc
    await self._ws.send(outgoing)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 635, in send
    await self.ensure_open()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 948, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: sent 1011 (internal error) keepalive ping timeout; no close frame received
websockets.exceptions.ConnectionClosedError: sent 1011 (internal error) keepalive ping timeout; no close frame received
------------------------------ Captured log call -------------------------------
WARNING  juju.client.connection:connection.py:654 RPC: Connection closed, reconnecting
WARNING  juju.client.connection:connection.py:654 RPC: Connection closed, reconnecting
ERROR    juju.client.connection:connection.py:662 RPC: Automatic reconnect failed
ERROR    asyncio:base_events.py:1707 Task exception was never retrieved
```

### To Reproduce

So far only reproducible in the CI.

1. Submit a PR from `main` (HEAD is in e0dc929 atm)
2. Observe the bundle integration job
3. Usually only the v2 fails

### Environment

CI environment
microk8s 1.29-strict/stable
juju 3.4 (3.4.4)
charmcraft latest/candidate (3.1.1)

### Relevant Log Output

```shell
See description, but can also be found [here](https://github.com/canonical/kfp-operators/actions/runs/10110429488/job/27960326357?pr=548#step:5:23)
```

**Comment by DnPlas** [comment]
I have tested the CI just running the v2 integration tests with the `--no-build` flag.

```
sg snap_microk8s -c "tox -e bundle-integration-${{ matrix.sdk }} -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2 --no-build"
```

This allows us to just see what's happening with the failing job (bundle-integration-v2) and at the same time, it confirmed that the issue was not at build time. With these changes, I could replicate the issue, and also observed the following:

```
  File "/home/runner/work/kfp-operators/kfp-operators/tests/integration/test_kfp_functional_v2.py", line 237, in test_viz_server_healthcheck
    status = await ops_test.model.get_status()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/model.py", line 2469, in get_status
    return await client_facade.FullStatus(patterns=filters)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/facade.py", line 486, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/_client6.py", line 1340, in FullStatus
    reply = await self.rpc(msg)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/facade.py", line 659, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/connection.py", line 649, in rpc
    await self._ws.send(outgoing)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 635, in send
    await self.ensure_open()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 939, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: sent 1011 (internal error) keepalive ping timeout; no close frame received
```

1. From logs [L156](https://github.com/canonical/kfp-operators/actions/runs/10148083446/job/28059985289?pr=548#step:5:157) shows a call to `status = await ops_test.model.get_status()`. This comes from [this line](https://github.com/canonical/kfp-operators/blob/main/tests/integration/test_kfp_functional_v2.py#L237) in the integration test file.
2. From logs [L162](https://github.com/canonical/kfp-operators/actions/runs/10148083446/job/28059985289?pr=548#step:5:163) then we see a call to `await self.rpc(msg)`
3. In logs [L166](https://github.com/canonical/kfp-operators/actions/runs/10148083446/job/28059985289?pr=548#step:5:167) we see how websockets is used `await self._ws.send(outgoing)`. This comes from [this line](https://github.com/juju/python-libjuju/blob/main/juju/client/connection.py#L6490). 
4. In `connection.py`, we can see the conditions under which the message `WARNING  juju.client.connection:connection.py:654 RPC: Connection closed, reconnecting` is logged ([L648-663](https://github.com/juju/python-libjuju/blob/main/juju/client/connection.py#L648C1-L663C26)):

```
            try:
                await self._ws.send(outgoing)
                break
            except websockets.ConnectionClosed:
                if attempt == 2:
                    raise
                log.warning('RPC: Connection closed, reconnecting')
                # the reconnect has to be done in a separate task because,
                # if it is triggered by the pinger, then this RPC call will
                # be cancelled when the pinger is cancelled by the reconnect,
                # and we don't want the reconnect to be aborted halfway through
                await jasyncio.wait([self.reconnect()])
                if self.monitor.status != Monitor.CONNECTED:
                    # reconnect failed; abort and shutdown
                    log.error('RPC: Automatic reconnect failed')
                    raise
```
5. From the logs, in [L171](https://github.com/canonical/kfp-operators/actions/runs/10148083446/job/28059985289?pr=548#step:5:172) we see that the websockets call failed `websockets.exceptions.ConnectionClosedError: sent 1011 (internal error) keepalive ping timeout; no close frame received`, which explains the error log from (4)

**Comment by DnPlas** [comment]
From the previous observation (https://github.com/canonical/kfp-operators/issues/549#issuecomment-2256490424), I tried two other things to avoid calling `await ops_test.model.get_status()` and avoid (potentially) the issue:

1. Use `url = await ops_test.model.applications["kfp-viz"].units[0].get_public_address()` instead.

As we can see from the [CI logs](https://github.com/canonical/kfp-operators/actions/runs/10149285336/job/28063873445#step:5:155), this is still causing an issue. Very similar issue from the one reported:

```
  File "/home/runner/work/kfp-operators/kfp-operators/tests/integration/test_kfp_functional_v2.py", line 239, in test_viz_server_healthcheck
    url = await ops_test.model.applications["kfp-viz"].units[0].get_public_address()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/unit.py", line 132, in get_public_address
    defResult = await app_facade.UnitsInfo(entities=[client.Entity(self.tag)])
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/facade.py", line 486, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/_client19.py", line 1763, in UnitsInfo
    reply = await self.rpc(msg)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/facade.py", line 659, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/juju/client/connection.py", line 649, in rpc
    await self._ws.send(outgoing)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 635, in send
    await self.ensure_open()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/websockets/legacy/protocol.py", line 948, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: sent 1011 (internal error) keepalive ping timeout; no close frame received
------------------------------ Captured log call -------------------------------
WARNING  juju.client.connection:connection.py:654 RPC: Connection closed, reconnecting
```

2. Use `json.loads(sh.juju.status(format="json", no_color=True, model="kubeflow"))["applications"]["kfp-viz"]["units"]["kfp-viz/0"]["address"]`

This way [actually works](https://github.com/canonical/kfp-operators/actions/runs/10149285336/job/28063873445#step:5:70). I assume it is because we are not calling anything that uses `websockets` under the hood (i.e. `get_status` or `get_public_address`).

PLEASE NOTE this does not explain why for v1 everything is passing and for v2 we get the error. For example, in [this](https://github.com/canonical/kfp-operators/actions/runs/10099006278/job/27957719480) run, the exact same code that executes the integration tests is passing for v1, but v2 is failing.

#### Workaround

To help alleviate this issue and unblock the CI, I suggest we apply the following change:

```diff
 async def test_viz_server_healthcheck(ops_test: OpsTest):
     """Run a healthcheck on the server endpoint."""
-    # unit = ops_test.model.units.get("kfp-viz/0")
-    # url = unit.data["public-address"]
-    url = await ops_test.model.applications["kfp-viz"].units[0].get_public_address()
+    url = json.loads(sh.juju.status(format="json", no_color=True, model="kubeflow"))["applications"]["kfp-viz"]["units"]["kfp-viz/0"]["address"]
```

**Comment by DnPlas** [comment]
I have found https://github.com/juju/python-libjuju/issues/458, which details the same issue as the one we are hitting in this repository. After checking with the juju team, we can confirm it is the same.

https://github.com/juju/python-libjuju/issues/458#issuecomment-754220158 explains in more detail the reason(s) for this disconnection to happen.

The juju team has also confirmed this is a known issue and that for now there is a workaround of setting juju's `MAX_FRAME_SIZE` to something bigger than the default. From [this conversation](https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$8EQFkaYz2AjNkH-Ezf7a-VhYQTQRxd3s1D4-7pk2dfk?via=ubuntu.com&via=matrix.org&via=mx.aouss.it):

> ...maybe you could try incresing the max frame size [here](https://github.com/juju/python-libjuju/blob/6c6d70d483937cec4404116c376b926106e05835/juju/client/connection.py#L241)? Its default is 4MB, but I believe this can occasionally been too little if a lot is changing very fast in Juju, since juju communicates with pylibjuju by sending a delta of all the updates since the last checkin.

> I'm not sure exactly what would cause it, but this is a known issue with pylibjuju. Pylibjuju has an internal state that it keeps in sync with the controller via the AllWatcher which sends deltas of the state. The juju CLI doesn't use this anymore because its a bit of an anti-pattern and work is currently underway to use alternatives in pylibjuju.

 This workaround is being tested in #553.

**Comment by kimwnasptd** [comment]
Excellent debug work!!
